### ISSUE TITLE
[feature request] Support HTTP proxy for tls_certificate data source

### ISSUE THREAD

**Issue by donovanmuller** [hashicorp/terraform-provider-tls#96]
I am trying to retrieve a certificate via the `tls_certificate` data source as per the [example in the docs](https://registry.terraform.io/providers/hashicorp/tls/latest/docs/data-sources/tls_certificate#example-usage). This is in context of an OIDC provider for an EKS cluster running in a private VPC (_no public network connectivity_). 

Given an OIDC issuer URL such as `https://oidc.eks.eu-west-1.amazonaws.com/id/838DB18FA8EC26D87848FDEC1D50CFDC` the call evetually fails with:

```
Error: dial tcp 34.248.140.150:443: connect: connection timed out
```

This is because of the private VPC and the fact that the data source _does not support a HTTP proxy_ (there are unfortunately no EKS VPC [endpoints](https://docs.aws.amazon.com/vpc/latest/userguide/integrated-services-vpce-list.html) currently).

### Terraform Version
`0.14.4`

### Affected Data Source(s)
- `tls_certificate`

### Terraform Configuration Files
```hcl
data "tls_certificate" "eks_cluster_oidc_issuer" {
  // E.g. https://oidc.eks.eu-west-1.amazonaws.com/id/838DB18FA8EC26D87848FDEC1D50CFDC
  url = aws_eks_cluster.eks_cluster.identity[0].oidc[0].issuer
}
```

### Expected Behavior
The `url` should support using a HTTP proxy. Currently https://github.com/hashicorp/terraform-provider-tls/blob/0fe77c596df5141b7f44b20643f92b439baccf63/internal/provider/data_source_tls_certificate.go#L92 does not.

### Actual Behavior
Connection eventually times out (`connection timed out`)

### References

- Perhaps adjusting/replacing the dialer with something similar to https://github.com/mwitkow/go-http-dialer would add HTTP proxy support

**Comment by rastakajakwanna** [comment]
Our team is trying to replace our custom data "external" containing shell script with this data source but we are blocked by exactly this issue. Proxy is a show stopper for this data source. I would expect every Terraform part to be following and supporting global proxy configuration which is not the case here.

Therefore, you have my +1, I was going to raise the same issue.

**Comment by hcabot** [comment]
I can not second that issue enough. I was struggling with that exact issue for a few days when a colleague pointed me here.
Glad it is now reported.

**Comment by alexmnyc** [comment]
same here. cannot be used for corporate networks

**Comment by guidoffm** [comment]
We also need this fix. Please implement it.

**Comment by guidoffm** [comment]
Is there someone who can approve https://github.com/hashicorp/terraform-provider-tls/pull/113

**Comment by liger1978** [comment]
If anyone else is having this issue, I have [forked this repo](https://github.com/liger1978/terraform-provider-tls), merged in [the fix](https://github.com/hashicorp/terraform-provider-tls/pull/113) from @DanielRis and [published a new version of the provider](https://registry.terraform.io/providers/liger1978/tls).  You can use this until Hashicorp do the needful:

```
terraform {
  required_providers {
    tls = {
      source = "liger1978/tls"
      version = "3.1.1"
    }
  }
}
```

**Comment by guidoffm** [comment]
> If anyone else is having this issue, I have [forked this repo](https://github.com/liger1978/terraform-provider-tls), merged in [the fix](https://github.com/hashicorp/terraform-provider-tls/pull/113) from @DanielRis and [published a new version of the provider](https://registry.terraform.io/providers/liger1978/tls). You can use this until Hashicorp do the needful:
> 
> ```
> terraform {
>   required_providers {
>     tls = {
>       source = "liger1978/tls"
>       version = "3.1.1"
>     }
>   }
> }
> ```

Maybe you could open a new PR for this.

**Comment by liger1978** [comment]
@guidoffm I don't see the point. @DanielRis's PR is still valid AFAICS.

**Comment by guidoffm** [comment]
> @guidoffm I don't see the point. @DanielRis's PR is still valid AFAICS.

The problem, as far as I remember, is, that DanielRis missed to sign the licence agreement and so the PR will stuck forever.
Maybe you can sign it and I hope to see progress then.

**Comment by liger1978** [comment]
@guidoffm I don't think I can be bothered with all that.

**Comment by guidoffm** [comment]
> @guidoffm I don't think I can be bothered with all that.

Don't you think it would work if you just create a new PR for your repo? Could you give it a try?

**Comment by DanielRis** [comment]
I will ping my HC account manager about this :)

**Comment by DanielRis** [comment]
Ok this has been fixed, sorry for that. I should have read better.

**Comment by liger1978** [comment]
@DanielRis No, your PR has still not been merged by HC: https://github.com/hashicorp/terraform-provider-tls/pull/113

**Comment by rajcheval** [comment]
We need this as well. I created a private EKS cluster. I am running in a bastion host but traffic is forced via corp proxy. Even after adding no-proxy environment variable we cannot fetch the tls_cert.

data "tls_certificate" "eks_cluster_oidc_issuer" {
  // E.g. https://oidc.eks.eu-west-1.amazonaws.com/id/838DB18FA8EC26D87848FDEC1D50CFDC
  url = aws_eks_cluster.eks_cluster.identity[0].oidc[0].issuer
}

**Comment by mattduguid** [comment]
thanks @liger1978 workaround sorted it for us, we'll keep that until this is merged then revert back

**Comment by adelwin** [comment]
i think this has regressed.

I'm using version 3.4.0 of this provider (latest).
I'm using a local CNTLM proxy (upstream proxy requires authentication, so I funnel it all via local CNTLM proxy).

Simple local tf setup.
```hcl
terraform {
  required_version = "~> 1.0"
  required_providers {
    tls = {
      # source  = "liger1978/tls"
      # version = "3.1.1"
      source  = "hashicorp/tls"
      version = "3.4.0"
    }
  }
}

data "tls_certificate" "this" {
  url = "https://oidc.eks.ap-southeast-1.amazonaws.com/id/"
  # url = "https://www.google.com"
}

output "fingerprint" {
  value = data.tls_certificate.this.certificates[0].sha1_fingerprint
}
```

A few things I've checked locally,
* I can reach the OIDC URL locally, using `curl -vvv`
* My proxy (both local and upstream) doesn't hijack the cert, and the cert returned are still AWS EKS cert (`subject: CN=oidc.eks.ap-southeast-1.amazonaws.com`)
* I've tried using www.google.com, still same result
* I've tried the same `main.tf` in a wide-open macbook it works fine, i.e. no proxy setup
